src/main/java/com/google/cloud/tools/appengine/AppEngineDescriptor.java

```diff
@@ -117,18 +117,16 @@ public boolean isJava8() throws AppEngineException {
    * </env-variables>
    * }</pre>
    *
-   * <p>This will construct a map of the form {[key, val], ...}.
+   * <p>This will construct a map of the form {[key, value], ...}.
    *
    * @return a map representing the environment variable settings in the appengine-web.xml
    */
-  // todo this should probably return an empty map instead per Effective Java
-  @Nullable
   public Map<String, String> getEnvironment() throws AppEngineException {
     Node environmentParentNode = getNode(document, "appengine-web-app", "env-variables");
     if (environmentParentNode != null) {
       return getAttributeMap(environmentParentNode, "env-var", "name", "value");
     }
-    return null;
+    return new HashMap<>();
```

Comment:
Collections.emptyMap()

Reply:
I don't think we should have a method that sometimes returns a mutable map and sometimes returns an immutable map depending on which code path is taken. furshlugginer Collections API :-(